File: src/main/scala/com/vc/hard/L212.scala
package com.vc.hard

object Solution {
  def findWords(board: Array[Array[Char]], words: Array[String]): List[String] = {
    import scala.collection.mutable
    val res = new mutable.HashSet[String]()

    case class Node(child: mutable.HashMap[Char, Node], var isLeaf: Boolean)
    class Trie {
      val root = new Node(new mutable.HashMap[Char, Node], false)
      def addWord(str: String): Unit = {
        var current = root
        str.indices.foreach(i => {
          val ch = str(i)
          if(!current.child.contains(ch)) {
            val node = new Node(new mutable.HashMap[Char, Node], false)
            current.child.put(ch, node)
          }
          current = current.child(ch)
        })
        current.isLeaf = true
      }

      def search(str: String): (Boolean, Boolean) = {
        var current = root
        str.indices.foreach(i => {
          val ch = str(i)
          if(!current.child.contains(ch)) return (false, false)
          current = current.child(ch)
        })
        (true, current.isLeaf)
      }
    }

    val n = board.length
    if(n == 0) return res.toList
    val m = board(0).length

    val visited = new mutable.HashSet[(Int, Int)]()
    val dirs =  Array((1, 0),(0, 1), (0, -1), (-1, 0))
    val trie = new Trie()
    def solve(x: Int, y: Int, str: String): Unit = {
      if(!visited.contains((x, y))) {
        visited += ((x, y))

        val searchRes = trie.search(str)
        if(!searchRes._1) {
          visited -= ((x, y))
          return
        }
        else if(searchRes._2) res += str

        dirs.foreach(dir => {
          val xNew = x + dir._1
          val yNew = y + dir._2
          if(xNew >= 0 && xNew < n && yNew >= 0 && yNew < m) {
            solve(xNew, yNew, str + board(xNew)(yNew))
          }
        })
        visited -= ((x, y))
      }
    }

    words.foreach(word => {
      trie.addWord(word)
    })
    board.indices.foreach(i => {
      board(i).indices.foreach(j => {
        visited.clear()
        solve(i, j, board(i)(j)+"")
      })
    })
    res.toList
  }
}

File: src/main/java/com/vc/hard/L212.java
package com.vc.hard;

import java.util.*;

class Node212 {
    boolean isLeaf;
    HashMap<Character, Node212> map;

    Node212() {
        isLeaf = false;
        map = new HashMap<Character, Node212>();
    }
}

class Trie212 {
    Node212 root = new Node212();
    public void addWord(String word) {
        Node212 current = root;
        for(int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            if(!current.map.containsKey(ch)) current.map.put(ch, new Node212());
            current = current.map.get(ch);
        }
        current.isLeaf = true;
    }

    public boolean[] search(String word) {
        Node212 current = root;
        for(int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            if(current.map.containsKey(ch)) current = current.map.get(ch);
            else return new boolean[]{false, false};
        }
        return new boolean[]{true, current.isLeaf};
    }
}

class Solution {
    Trie212 trie = new Trie212();

    public List<String> findWords(char[][] board, String[] words) {
        int n = board.length;
        int m = board[0].length;
        for(String word: words) trie.addWord(word);

        HashSet<String> res = new HashSet<String>();
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                boolean[][] visited = new boolean[n][m];
                solve(i, j, board, new StringBuilder(), visited, res, n, m);
            }
        }
        return new ArrayList<String>(res);
    }

    int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    private void solve(int row, int col, char[][] board,
                       StringBuilder str, boolean[][] visited,
                       HashSet<String> res,
                       int n, int m) {
        if(!visited[row][col]) {
            visited[row][col] = true;
            str.append(board[row][col]);
            boolean[] search = trie.search(str.toString());

            if(!search[0]) {
                visited[row][col] = false;
                str.setLength(str.length() - 1);
                return;
            }
            else if(search[1]) res.add(str.toString());

            for(int[] dir: dirs) {
                int rowNew = row + dir[0];
                int colNew = col + dir[1];
                if(rowNew >= 0 && rowNew < n && colNew >= 0 && colNew < m) {
                    solve(rowNew, colNew, board, str, visited, res, n, m);
                }
            }
            visited[row][col] = false;
            str.setLength(str.length() - 1);
        }
    }
}
